### DISCUSSION TITLE
idleTimeout property doesn't work for values less than 6 seconds

### DISCUSSION THREAD

**Discussion by alexander-torosh** [uNetworking/uWebSockets.js#396] [Q&A] [answered]
Hello @alexhultman 
I found the next issue in the app behavior:
- Browser client sends empty WS messages to the server every second for keeping the connection alive.
- The server receives these empty messages from the browser and does nothing.
- If I set the application `idleTimeout` property from 1 to 5, the server detects WS disconnection with code `1006` every 1-5 seconds after the client connection.
- If I set the `idleTimeout` property value as 6 or greater, the server properly handles the idle logic.

Seem like it works with values that greater than 5.
Thank you!

**Comment by ghost** [comment]
The granularity of timeouts is 4 seconds

**Comment by alexander-torosh** [comment]
It means that `idleTimeout` property would work with values >= 6, correct?
If yes, then it is not clear and the documentation could be improved there. 
Thanks

**Comment by hst-m** [comment, marked as answer]
@alexander-torosh as described in this issue https://github.com/uNetworking/uWebSockets.js/issues/286#issuecomment-608627878 the idleTimeout number rounds to the nearest 4 seconds, so setting idleTimeout to anywhere from 2 and less than 6 rounds to 4. Anywhere from 6 to less than 10 rounds to 8, etc. If the client is idle anywhere from that number and 4 less than that number it could trigger the `close` event. So idleTimeout of 4 is bad because any client idle time between 0 and 4 could trigger the `close` event so it closes immediately. You need at least idleTimeout of 8 to work, where any idle client time from 4 to 8 seconds could trigger the `close` event. In that case you want to send Ping messages at least every 3 seconds. I use idleTimeout of 16, which rounds to 16, then any idle client time between 12 and 16 triggers the `close` event, and I have the client sending "ping" empty ws messages every 10 seconds to keep it open. I reset the timer if the client sends or receives any messages

**Comment by hst-m** [comment]
I think it might make more sense if the idleTimeout is changed to indicate an idle client time of between that number and Plus 4 seconds instead of Minus 4 seconds. So setting an idleTimeout of 4 seconds means idle client from 4 to 8 seconds gets closed, and idleTimeout of 8 means from 8 to 12 seconds gets closed. Because it usually uses the lower range to close the connection, which is 4 seconds less than what people think they are getting when setting idleTimeout. As long as people understand what is going on it doesn't matter but it might be easier if is changed or better documented

**Comment by ghost** [comment]
I don't remember the exact rules here but I'm pretty sure this is correct. Timeouts are not meant to be accurate.
